ember-toucan-core/src/components/form/file-input-field.ts

```diff
@@ -0,0 +1,74 @@
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { assert } from '@ember/debug';
+import { action } from '@ember/object';
+
+export type OnFileUploadChangeCallback = (
+  files: FileList,
+  e: Event | InputEvent
+) => void;
+
+interface ToucanFormFileInputFieldComponentSignature {
+  Element: HTMLInputElement;
+  Args: {
+    accept?: string;
+    error?: string;
+    files?: FileList;
+    hint?: string;
+    label: string;
+    isDisabled?: boolean;
+    multiple?: boolean;
+    onChange?: OnFileUploadChangeCallback;
+    rootTestSelector?: string;
+  };
+  Blocks: {
+    triggerText: [];
+  };
+}
+
+export default class ToucanFormFileInputFieldComponent extends Component<ToucanFormFileInputFieldComponentSignature> {
+  @tracked files;
+
+  constructor(
+    owner: unknown,
+    args: ToucanFormFileInputFieldComponentSignature['Args']
+  ) {
+    assert(
+      'A "@label" argument is required for FileInputField',
+      args.label !== undefined
+    );
+    super(owner, args);
+
+    if (args.files) {
+      this.files = args.files;
+    }
+  }
+
+  get accept() {
```

Comment:
Props/Arg question: I am conflicted about mixing args and @tracked values like this (in React it's considered a bit [of a code smell](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#anti-pattern-unconditionally-copying-props-to-state))

This pattern does allow default @args however (for `accepts` and `multiple`). 

Is there a better way to handle default @args yet? I remember Preston saying there wasn't...

Reply:
No, I think it's fine!

I had only a quick look at your React document, and also I don't know a lot about React, but I don't see _any_ problem here (except for maybe having to type this here in a verbose way, so something for "default value for args" was [discussed](https://github.com/emberjs/rfcs/issues/479) a while ago, doesn't seem it will come tho).

> I am conflicted about mixing args and @Tracked values like this

But you are _not_ doing this! There is no `@tracked` here. The getter is fully derived state, nothing except for the actuals arg is tracked here, so also nothing can get out of sync or diverge in any way. 

It would be different if you did this instead:

```js
@tracked accept = this.args.accept ?? '*';
```

Then indeed you would have two copies of `accept`. But you are not actually overriding the arg with some local state or so. Whenever `this.arg.accept` changes, `this.accept` will automatically reflect that. In that sense, I think what we have here is basically an [controlled component](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-controlled-component) as defined in your React doc.